### PR TITLE
[Turbo] Support custom TurboStreamResponse actions

### PR DESCRIPTION
**Pull request by DRaichev**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| Issues        | -
| License       | MIT

The current implementation of TurboStream & TurboStreamResponse has no generic action (only the predefined turbo actions) and what's more the class is marked as final, which does not allow adding support for more actions. I think the wrap function should be public or have a public function "custom" that exposes the functionality (I've opted for the latter)

I really like this library: https://github.com/marcoroth/turbo_power
It adds additional simple actions. For example I use set/delete attributes to enable/disable UI elements

This change will allow people to use this or similar libraries, or even build their own custom actions if they want to.

There is no impact to the rest of the functionality or DX

There is a minor change to the wrap function, it now accepts an array of attributes instead of a string. It builds the string from the array of attributes, and adds a leading space. This aims to prevent errors, as the previous implementation needs the $attr argument to start with a blank space otherwise it would produce invalid html

### CODE REVIEW

**Comment by setineos** [comment]
I think it should also work for `<twig:Turbo:Stream:*>` components.

File: src/Turbo/src/Helper/TurboStream.php [hunk 1]
```diff
@@ -86,12 +86,28 @@ public static function refresh(?string $requestId = null): string
...
+     */
+    public static function custom(string $action, string $target, string $html, array $attr = []): string
     {
+        return self::wrap($action, $target, $html, $attr);
+    }
+
+    /**
+     * @param array<string> $attr
+     */
+    private static function wrap(string $action, string $target, string $html, array $attr = []): string
```
Comment (nicolas-grekas) [inline src/Turbo/src/Helper/TurboStream.php]: we don't need that method anymore, custom is basically a public version of wrap

File: src/Turbo/src/Helper/TurboStream.php [hunk 2]
```diff
@@ -86,12 +86,28 @@ public static function refresh(?string $requestId = null): string
...
+    {
+        // Join array elements with a space and prepend a leading space
+        $atrrString = empty($attr) ? '' : ' '.implode(' ', $attr);
```
Comment (nicolas-grekas) [inline src/Turbo/src/Helper/TurboStream.php]: the API for $attr means people are responsible for encoding their attributes, is that what we want?

Comment (nicolas-grekas) [reply src/Turbo/src/Helper/TurboStream.php]:
```suggestion
        $atrrString = $attr ? ' '.implode(' ', $attr) : '';
```

**Comment by nicolas-grekas** [comment]
> I think it should also work for <twig:Turbo:Stream:*> components.

This can be done by adding a generic `<twig:Turbo:Stream>` component maybe.

File: src/Turbo/src/Helper/TurboStream.php (hunk 2 above)
Comment (DRaichev) [reply src/Turbo/src/Helper/TurboStream.php]: > the API for $attr means people are responsible for encoding their attributes, is that what we want?

Not sure, I don't mind either way, but I think it should be documented if users are responsible for encoding.
What do you think ?

Comment (DRaichev) [reply src/Turbo/src/Helper/TurboStream.php]: >  $atrrString = $attr ? ' '.implode(' ', $attr) : '';

I'm not sure I like this. I generally don't like implicit checks, although for empty array it makes sense.

Comment (nicolas-grekas) [reply src/Turbo/src/Helper/TurboStream.php]: I think I'd prefer encoding on our side, and $attr be a name=>value list
about the check, empty() is the worst :)

**Comment by DRaichev** [comment]
@nicolas-grekas  
I just realised if you pass "action" or "targets" into the attributes array those will conflict with the existing attributes.
I think we should throw an exception if those are present

**Comment by smnandre** [comment]
With the incoming [`<twig:Turbo:Frame>`](https://github.com/symfony/ux/pull/2303) and [`<twig:Turbo:Stream>`](https://github.com/symfony/ux/pull/2302), do you think we still need this features ?

File: src/Turbo/src/Helper/TurboStream.php [hunk 3]
```diff
@@ -61,15 +61,15 @@ public static function remove(string $target): string
...
+        return self::custom('before', $target, $html);
     }
 
     /**
      * Inserts after the element(s) designated by the target CSS selector.
      */
     public static function after(string $target, string $html): string
     {
-        return self::wrap('after', $target, $html);
+        return self::custom('after', $target, $html);
```
Comment (smnandre) [inline src/Turbo/src/Helper/TurboStream.php]: I think we should revert these changes (other ones too in existing methods)

(wrap worked better as method name here)

File: src/Turbo/src/Helper/TurboStream.php [hunk 4]
```diff
@@ -86,12 +86,34 @@ public static function refresh(?string $requestId = null): string
         return \sprintf('<turbo-stream action="refresh" request-id="%s"></turbo-stream>', htmlspecialchars($requestId));
     }
 
-    private static function wrap(string $action, string $target, string $html, string $attr = ''): string
+    /**
+     * Custom action and attributes.
+     *
+     * @param array<string, string|int|null> $attr
+     */
+    public static function custom(string $action, string $target, string $html, array $attr = []): string
```
Comment (smnandre) [inline src/Turbo/src/Helper/TurboStream.php]: I'd rather keep the wrap method, and have this new method use it 

(prepare and escape the $attr string and pass it to self::wrap())

File: src/Turbo/src/Helper/TurboStream.php [hunk 5]
```diff
@@ -86,12 +86,34 @@ public static function refresh(?string $requestId = null): string
...
     {
+        if (\array_key_exists('action', $attr) || \array_key_exists('targets', $attr)) {
+            throw new \InvalidArgumentException('The "action" and "targets" attributes are reserved and cannot be used.');
+        }
+
+        $attr['action'] = $action;
+        $attr['targets'] = $target;
+        $attr = array_merge(['action' => $action, 'targets' => $target], $attr);
```
Comment (smnandre) [inline src/Turbo/src/Helper/TurboStream.php]: Not needed if you call the `wrap` method

File: src/Turbo/src/Helper/TurboStream.php [hunk 6]
```diff
@@ -86,12 +86,34 @@ public static function refresh(?string $requestId = null): string
...
+
+        $attrString = '';
+        foreach ($attr as $key => $value) {
+            if (null === $value) {
+                $attrString .= \sprintf(' %s', $key);
+            } else {
+                $attrString .= \sprintf(' %s="%s"', $key, \is_string($value) ? htmlspecialchars($value) : $value);
```
Comment (smnandre) [inline src/Turbo/src/Helper/TurboStream.php]: If $value is a Stringable object, this can be dangerous.

Special cases for boolean should be added too probably.

File: src/Turbo/src/Helper/TurboStream.php [hunk 7]
```diff
@@ -86,12 +86,34 @@ public static function refresh(?string $requestId = null): string
...
+        }
+
+        $attr['action'] = $action;
+        $attr['targets'] = $target;
+        $attr = array_merge(['action' => $action, 'targets' => $target], $attr);
+
+        $attrString = '';
+        foreach ($attr as $key => $value) {
+            if (null === $value) {
+                $attrString .= \sprintf(' %s', $key);
```
Comment (smnandre) [inline src/Turbo/src/Helper/TurboStream.php:106]: Don't we need to encode keys here?

File: src/Turbo/src/TurboStreamResponse.php [hunk 8]
```diff
@@ -104,4 +104,16 @@ public function refresh(?string $requestId = null): static
 
         return $this;
     }
+
+    /**
+     * @param array<string> $attr
+     *
+     * @return $this
+     */
+    public function custom(string $action, string $target, string $html, array $attr = []): static
```
Comment (smnandre) [inline src/Turbo/src/TurboStreamResponse.php]:
```suggestion
    public function action(string $action, string $target, string $html, array $attr = []): static
```

?

File: src/Turbo/src/TurboStreamResponse.php [hunk 9]
```diff
@@ -104,4 +104,16 @@ public function refresh(?string $requestId = null): static
 
         return $this;
     }
+
+    /**
+     * @param array<string> $attr
```
Comment (smnandre) [inline src/Turbo/src/TurboStreamResponse.php]:
```suggestion
     * @param array<string, scalar> $attr
```

?

**Comment by DRaichev** [comment]
> With the incoming [`<twig:Turbo:Frame>`](https://github.com/symfony/ux/pull/2303) and [`<twig:Turbo:Stream>`](https://github.com/symfony/ux/pull/2302), do you think we still need this features ?

I think they are complementary. I'd rather use the twig component syntax only in templates and stick to this helper on the php side, plus it's better to have consistent functionality in both, especially for people who dislike the twig component syntax.

File: src/Turbo/src/TurboStreamResponse.php (hunk 9 above)
Comment (DRaichev) [reply src/Turbo/src/TurboStreamResponse.php]: Scalar includes bool, and I think passing null as value for boolean attributes is better, as per
https://developer.mozilla.org/en-US/docs/Glossary/Boolean/HTML

So I think `array<string, string|int|float|null>` is the correct type.

File: src/Turbo/src/Helper/TurboStream.php (hunk 7 above)
Comment (DRaichev) [reply src/Turbo/src/Helper/TurboStream.php:106]: I don't really see a case where the key would have special characters, but it doesn't hurt, so better safe than sorry

File: src/Turbo/src/Helper/TurboStream.php (hunk 6 above)
Comment (DRaichev) [reply src/Turbo/src/Helper/TurboStream.php]: I've changed it so it covers the types specified in the typehint and would fail if an object is passed

**Comment by smnandre** [comment]
Looks good to me!

Could you check the type of values and throw if not expected (a non-stringable object) ?

**Comment by DRaichev** [comment]
> Looks good to me!
> 
> Could you check the type of values and throw if not expected (a non-stringable object) ?

I believe this is going to throw:
```
foreach ($attr as $key => $value) {
    $key = htmlspecialchars($key);
    if (null === $value) {
        $attrString .= \sprintf(' %s', $key);
    } elseif (\is_int($value) || \is_float($value)) {
        $attrString .= \sprintf(' %s="%s"', $key, $value);
    } else {
        $attrString .= \sprintf(' %s="%s"', $key, htmlspecialchars($value)); // htmlspecialchars expect string so will throw here
    }
}
```

Or do you mean somewhere else ?

**Comment by DRaichev** [comment]
> > Looks good to me!
> > Could you check the type of values and throw if not expected (a non-stringable object) ?
> 
> I believe this is going to throw:
> 
> ```
> foreach ($attr as $key => $value) {
>     $key = htmlspecialchars($key);
>     if (null === $value) {
>         $attrString .= \sprintf(' %s', $key);
>     } elseif (\is_int($value) || \is_float($value)) {
>         $attrString .= \sprintf(' %s="%s"', $key, $value);
>     } else {
>         $attrString .= \sprintf(' %s="%s"', $key, htmlspecialchars($value)); // htmlspecialchars expect string so will throw here
>     }
> }
> ```
> 
> Or do you mean somewhere else ?

So this is ready for merge then, right

**Comment by smnandre** [comment]
After some thinking, why don't we rename the `custom($action)`  method in `action($name, .... ... ... .. )`  ?

**Comment by DRaichev** [comment]
> After some thinking, why don't we rename the `custom($action)` method in `action($name, .... ... ... .. )` ?

Yeah, I was thinking about this when I renamed the one in TurboStreamResponse
It should be either custom in both or action in both.

I don't really have a preference, both work, so I leave the choice to you

**Comment by smnandre** [comment]
Let's use action in both then (sorry for the late comment)

File: src/Turbo/src/Helper/TurboStream.php [hunk 10]
```diff
@@ -86,12 +86,45 @@ public static function refresh(?string $requestId = null): string
         return \sprintf('<turbo-stream action="refresh" request-id="%s"></turbo-stream>', htmlspecialchars($requestId));
     }
 
+    /**
+     * Custom action and attributes.
+     * When passing attributes use null value to for boolean attributes (e.g. disabled).
```
Comment (smnandre) [inline src/Turbo/src/Helper/TurboStream.php]:
```suggestion
     * 
     * Set boolean attributes (e.g., `disabled`) by providing the attribute name as key with `null` as value.
```

File: src/Turbo/src/Helper/TurboStream.php [hunk 11]
```diff
@@ -86,12 +86,45 @@ public static function refresh(?string $requestId = null): string
...
             <turbo-stream action="%s" targets="%s"%s>
                 <template>%s</template>
             </turbo-stream>
-            EOHTML, $action, htmlspecialchars($target), $attr, $html);
+            EOHTML,
+            $action,
+            htmlspecialchars($target),
+            $attr,
+            $html
+        );
```
Comment (smnandre) [inline src/Turbo/src/Helper/TurboStream.php]: Let's revert this change, as the previous one feels more readable

**Review by smnandre — COMMENTED** [review]
Last couple of minor suggestions..

Could you update the CHANGELOG in the src/Turbo repository ? 

And let's merge this :)

**Comment by DRaichev** [comment]
Ready to merge

**Review by smnandre — APPROVED** [review]
Thank you @DRaichev !

File: src/Turbo/CHANGELOG.md [hunk 12]
```diff
@@ -1,4 +1,7 @@
 # CHANGELOG
+## 2.22.0
+
+-   Add support for custom actions in `TurboStream` and `TurboStreamResponse`
```
Comment (smnandre) [inline src/Turbo/CHANGELOG.md]: Arf, saw this just too late... can you move it a couple lines down, there was already a 2.22.0 block 😅

Comment (DRaichev) [reply src/Turbo/CHANGELOG.md]: Ah, yes, it changed when synced with the upstream.

**Comment by DRaichev** [comment]
Updated the changelog, now ready to merge

**Review by smnandre — APPROVED** [review]
Perfect!

**Comment by DRaichev** [comment]
@smnandre Hey, can we get this merged ?

I am in the process of replacing the usage of the old syntax at work. I've already replaced all basic cases, so 
now I'm waiting for v2.22 to do the `<twig:Turbo:Frame/>` and hopefully we can get this as well so I can do the instances of custom turbo actions.

**Comment by Kocal** [comment]
Hey, I rebasing your PR and will merge it in the next minutes.

**Comment by Kocal** [comment]
Thank you @DRaichev.

**Comment by smnandre** [comment]
@DRaichev we will probably release a 2.22 in the coming weeks, you can use it with the "2.x-dev" in the meanwhile! Thanks for this PR 😄

**Comment by DRaichev** [comment]
@Kocal  @smnandre Thanks for the speedy reaction. And thank you for all the work you're doing on this project. 
I used to hate to work on UX stuff, but thanks to Turbo and Symfony UX I realised you can have great UX without the convoluted mess. Keep up the good work.

**Comment by smnandre** [comment]
Thank you for such a kind message—it really means a lot!
